substrate/runtime/staking/src/lib.rs

```diff
@@ -818,27 +816,44 @@ impl<T: Trait> Module<T> {
 			return Err("bondage too high to send value");
 		}
 
-		let dest = T::DetermineContractAddress::contract_address_for(code, transactor);
+		let dest = T::DetermineContractAddress::contract_address_for(constructor_code, transactor);
 
 		// early-out if degenerate.
 		if &dest == transactor {
 			return Ok(None);
 		}
 
-		let mut local = BTreeMap::new();
-		// two inserts are safe
+		let mut overlay = OverlayAccountDb::new(account_db);
+
+		// update balances of the sender and the created account.
 		// note that we now know that `&dest != transactor` due to early-out before.
-		local.insert(dest, ChangeEntry::contract_created(value, code.to_vec()));
-		local.insert(transactor.clone(), ChangeEntry::balance_changed(from_balance - liability));
-		Ok(Some(local))
+		overlay.set_balance(&dest, value);
+		overlay.set_balance(transactor, from_balance - liability);
+
+		// TODO: an additional fee, based upon gaslimit/gasprice.
+		let gas_limit = 100_000;
+
+		let exec_result = {
+			let mut staking_ext = StakingExt {
+				account_db: &mut overlay,
+				account: dest.clone(),
+			};
+			contract::execute(constructor_code, &mut staking_ext, gas_limit)
+				.map_err(|_| "execution of contract constructor failed")?
```

Comment:
all changes are reverted, right? is that intended? gas is presumably still charged?